src/main/java/formflow/library/data/Submission.java

```diff
@@ -251,6 +272,60 @@ private static Map<String, Object> copyMap(Map<String, Object> origMap) {
     return result;
   }
 
+  /**
+   * getShortCode returns a read-only unique code for the submission after the submission has been saved to the database initially.
+   * The short code generation is configurable via @ShortCodeConfig. The length, characterset (alphanumeric, numeric, alpha), and
+   * forced uppercasing are available (with the defaults being 6, alphanumeric, and true respectively).
+   *
+   * This method will check if the generated code exists in the database, and keep trying to create a unique code, before
+   * persisting and returning the newly generated code-- therefore it is very important to ensure the configuration allows for a
+   * suitably large set of possible codes for the application.
+   * @return the short code for the Submission
+   */
+  public synchronized String getShortCode() {
+
+    if (this.createdAt == null) {
+      // If the submission hasn't been saved ever, there should be no concept of a short code
+      return null;
+    }
+
+    if (submissionRepository == null) {
+      submissionRepository = (SubmissionRepository) SpringConfiguration.contextProvider().getApplicationContext().getBean("submissionRepository");
+    }
+
+    shortCode = submissionRepository.findShortCodeById(this.id);
+    if (shortCode == null) {
+      if (shortCodeConfig == null) {
+        shortCodeConfig = (ShortCodeConfig) SpringConfiguration.contextProvider().getApplicationContext().getBean("shortCodeConfig");
+      }
+
+      // If there is no short code for this submission in the database, generate one
+      do {
+        int codeLength = shortCodeConfig.getCodeLength();
+        String newCode = switch(shortCodeConfig.getCodeType()) {
+          case alphanumeric -> RandomStringUtils.randomAlphanumeric(codeLength);
+          case alpha -> RandomStringUtils.randomAlphabetic(codeLength);
+          case numeric -> RandomStringUtils.randomNumeric(codeLength);
+        };
+
+        if (shortCodeConfig.isUppercase()) {
+          newCode = newCode.toUpperCase();
+        }
+
+        boolean exists = submissionRepository.existsByShortCode(newCode);
+        if (!exists) {
+          // If the newly generated code isn't already in the database being used by a prior submission
+          // set this submission's shortcode, and persist it
+          shortCode = newCode;
+          submissionRepository.save(this);
+        } else {
+          log.warn("Confirmation code {} already exists", newCode);
+        }
+      } while (shortCode == null);
+    }
+    return shortCode;
+  }
```

Comment:
Seeing this now I wonder if it makes sense to do this in the ScreenController on Post when you save the submission for the first time? 

That would avoid the need to have applicationContext and SubmissionRepository as dependencies here since the screen controller has SubmissionRepositoryService as a dependency already and you could then have ShortCodeConfig as a dependency of SubmissionRepositoryService.

That said, my Hibernate / JPA foo is not super strong, but my main concern with this is the ApplicationContext, since the library gets pulled in by other applications which will already have their own context. 

If we use this implementation it might be a good idea to locally pull this library branch into the starter-app and see if it plays nice there.

Reply:
@spokenbird this is all cleaned up now in the PR. I still have to write more tests, but functionally this should be what we discussed on Tuple!